qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,235 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A generalized QAOA quantum circuit with a support of custom initial states and mixers."""
+# pylint: disable=R0401
+from typing import Optional, Union, cast, Set
+
+import numpy as np
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.parametervector import ParameterVector
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from ..blueprintcircuit import BlueprintCircuit
+
+
+class QAOAAnsatz(BlueprintCircuit):
+    """A generalized QAOA quantum circuit with a support of custom initial states and mixers.
+
+    References:
+
+        [1]: Farhi et al., A Quantum Approximate Optimization Algorithm.
+            `arXiv:1411.4028 <https://arxiv.org/pdf/1411.4028>`_
+    """
+    def __init__(self,
+                 cost_operator: 'OperatorBase',
+                 reps: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, 'OperatorBase']] = None,
+                 name: str = "qaoa"):
+        r"""
+        Args:
+            cost_operator: The operator representing the cost of the optimization problem,
+                denoted as :math:`U(C, \gamma)` in the original paper.
+            reps: The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper.
+            initial_state: An optional initial state to use. If `None` is passed then a set of
+                Hadamard gates is applied as an initial state to all qubits.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as :math:`U(B, \beta)` in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._reps = reps
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = None
+        self._num_parameters = 0
+        self._bounds = None
+        self._mixer = None
+        self._data = None
+
+    def _check_configuration(self, raise_on_failure: bool = True) -> bool:
+        valid = True
+
+        if self._cost_operator is None:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The operator representing the cost of "
+                                     "the optimization problem is not set")
+        if self._reps is None or self._reps < 1:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The integer parameter p, which determines the depth "
```

Comment:
This should say `reps` instead of `p`. And would it not be simpler, and easier undertand if it were `f'reps needs to be >=1 but has value: {reps}'` So it gets rid of the dual condition in the explanation and includes whatever the failing value is.

Reply:
Agree.